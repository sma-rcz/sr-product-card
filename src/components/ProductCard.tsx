import React, { createContext } from 'react'
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks/useProduct'
import { InitialValues, onChangeArgs, Product, ProductCardHadnlers, ProductContextProps } from '../interfaces/interfaces';



// Contexto para compartir el estado del producto a sus hijos
export const ProductContext = createContext({} as ProductContextProps); // se inicializa con un objeto vacio y se le asigna el tipo de dato que va a tener
const { Provider } = ProductContext  // destructuring para obtener el provider





export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args:ProductCardHadnlers) => JSX.Element;
  className?: string; // se define el tipo de la propiedad para poder agrer stylos
  style?:React.CSSProperties; // se necesita para poder agregar estilos al componente React.CSSProperties
  onChange?: (args:onChangeArgs) => void; // se define el tipo de la propiedad para poder agrer stylos
  value?:number;
  initialValues?:InitialValues
}



// Componente que renderiza la tarjeta del producto
export const ProductCard = ({ children, product,className ,style, onChange ,value,initialValues}: Props) => {
  const { counter, increseBy, maxCount
    , isMaxCountReached, reset} = useProduct(
    {
      onChange, // se le pasa la funcion onChange al hook
      product, // se le pasa el producto al hook
      value,
      initialValues,
      
   

  });


  return (
    <>{/* se envuelve todo en el provider para que los hijos puedan acceder al estado del producto */}
      <Provider value={{

        counter,
        increseBy,
        product,
        maxCount,
      }}>
        <div 
        style={style}
        className={`${styles.productCard} ${className}`}>

          {/* <span className={styles.productDescription}>{product.title}</span>*/}

          {children({
            count:counter,
            isMaxCountReached,
            maxCount:maxCount,
            product,
            increseBy,
            reset
          })}



        </div>
      </Provider>

    </>

  )
}

