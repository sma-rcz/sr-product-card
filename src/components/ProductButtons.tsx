import React ,{ useCallback, useContext} from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css'


interface ProductButtonsProps {
  className?: string;  // aqui se define el tipo de la propiedad para poder agrer stylos
  style?:React.CSSProperties; // se necesita para poder agregar estilos al componente React.CSSProperties
}



                          // se le asigna el tipo de las propiedades que se van a recibir de los estilos
export const ProductButtons = ({className ,style}:ProductButtonsProps) => {
  


  //Todo:MaxCount
    const {counter,increseBy ,maxCount} = useContext(ProductContext); // aqui se obtiene el estado del producto
  //Todo: isMaxReached = useCallback, dependencias [counter, maxCount]
  //TRUE si el count  === maxCount
  //FALSE si el count !== maxCount  

  /*
  const isMaxReached  = useMemo (() => counter === maxCount,[counter,maxCount] ); // se crea un memo para saber si el contador es igual al maximo*/
  
  const isMaxReached =  useCallback(
    () => !!maxCount && counter === maxCount,[counter,maxCount] ); // se crea un memo para saber si el contador es igual al maximo

    return (
      <div
      style={style} // se le asigna el estilo al componente
      className={`${styles.buttonsContainer} ${className}`}> {/* aqui se asigna los stilos de Classname */}
        <button
         disabled = {counter <= 0} // se desabilita el boton si el contador es igual a 0
          onClick={() => increseBy(-1)}
          className={styles.buttonMinus}>-</button>
  
        <div className={styles.countLabel}>{counter}</div>
  
        <button
          disabled = {isMaxReached()} // se desabilita el boton si el contador es igual al maximo
          onClick={() => increseBy(+1)}
          className={`${styles.buttonAdd} ${isMaxReached() ? styles.disabled : '' } `} >+</button>
      </div>
      )
  
  }
  