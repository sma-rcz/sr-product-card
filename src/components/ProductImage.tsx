import React,{ useContext } from "react";
import { ProductContext } from "./ProductCard";
import NoImg from '../assets/no-image.jpg' // Ajusta la ruta segÃºn sea necesario
import styles from '../styles/styles.module.css'

// Componente que renderiza la imagen del producto desde una funcion  que recibe la url de la imagen

export interface ProductImageProps {
  img?: string | undefined;
  className?: string;
  style?:React.CSSProperties;
}




export const ProductImage = ({ img = '',className,style }:ProductImageProps ) => {

    const  {product  } = useContext(ProductContext);
    let imgToshow:string
    if( img){
      imgToshow = img
    }else if(product.img){
      imgToshow = product.img
    }
    else{
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      imgToshow = NoImg
    }
  
  
  
  
    return (<img 
      className={`${styles.productImg} ${className}`} src={imgToshow }  alt='Coffee mug' width='200px' height='200px'
      style={style}
      
      />)
  }

