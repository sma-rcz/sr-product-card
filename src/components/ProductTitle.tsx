import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css'


export interface ProductTitleProps {
  title?:string,
  className?:string,
  style?:React.CSSProperties
}



export const ProductTitle = ({ title,className,style }:ProductTitleProps ) => {

  
    const {product} = useContext(ProductContext);
    /*
    if(title  === ''){
      title = product.title
    }*/
   //title ? title === ' ' : product.title
    
    return (<span 
      style={style}
      className={`${styles.productDescription} ${className}`}>{ title = title === '' ? product.title : title}</span>)
  }