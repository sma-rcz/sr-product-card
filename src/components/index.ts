import { ProductCard as ProductCardHOC } from './ProductCard';


import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductCardHOCprops } from '../interfaces/interfaces';

//export{ ProductCard  as ProductCardHOC} from "./ProductCard";


//export{ ProductCard } from "./ProductCard";

export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";
export { ProductButtons } from "./ProductButtons";


// aqui se exporta el componente ProductCard con los elementos que se van a exportar todo esto es para los hijos de ProductCard
export const ProductCard:ProductCardHOCprops = Object.assign(ProductCardHOC, { // se le asigna a ProductCardHOC los elementos que se van a exportar
    Image: ProductImage,    
    Title: ProductTitle,
    Buttons: ProductButtons
});

export default ProductCard;
