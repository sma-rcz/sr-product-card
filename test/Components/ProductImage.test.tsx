import renderer from 'react-test-renderer';
import { ProductImage ,ProductCard } from '../../src/components';
import React from 'react';
import { product2 } from '../data/products';
describe('debe de mostrar el Componente de la img', () => { 

    test('debe de mostrar el component con el url de img', () => {
        const wrapper = renderer.create(
            <ProductImage img='img/imagen-personalizada.jpg'  />
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
      
    });

    test('debe de mostrar la imagen con el nombre   ', () => {
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    ()=>(
                        <ProductImage  />
                    )
                }


            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
      
    })
    
    
 })