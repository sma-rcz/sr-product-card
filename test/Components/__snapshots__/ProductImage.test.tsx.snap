// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debe de mostrar el Componente de la img debe de mostrar el component con el url de img 1`] = `
<img
  alt="Coffee mug"
  className="productImg undefined"
  height="200px"
  src="img/imagen-personalizada.jpg"
  width="200px"
/>
`;

exports[`debe de mostrar el Componente de la img debe de mostrar la imagen con el nombre    1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffee mug"
    className="productImg undefined"
    height="200px"
    src="../assets/no-image.jpg"
    width="200px"
  />
</div>
`;
